README.md
# <a name="top"></a>sAPI

* [概述](#概述)
* [安装](#安装)
* [Demo](#demo)
* [使用](#使用)
   * [Objc](#objc)
   * [Swift](#swift)
   * [自适应服务端返回数据结构](#adaptret)
* [公共参数](#公共参数)
   * [commonParams](#commonparams)
   * [commonHeaders](#commonheaders)
   * [区分不同子系统](#tags)

## 概述

业务无关的链式API调用工具，支持 objc / swift

底层网络访问基于 [AFNetworking](https://github.com/AFNetworking/AFNetworking)

异步返回处理基于google的 [promise](https://github.com/google/promises)

## 安装

通过pod引用，在podfile增加下面一行，通过tag指定版本
```
pod 'sAPI',         :git => "https://github.com/jinsikui/sAPI.git", :tag => 'v1.0.3-0'
```
 objc代码中引入：
```
#import <sAPI/sAPI.h>
```
 swift代码中引入：
```
在项目的plist文件中配置置 'Objective-C Bridging Header = xxx.h'，在xxx.h中添加一行：
#import <sAPI/sAPI.h>
```

## <a name="demo"></a> Demo

直接下载本工程运行即可启动demo

## 使用

### <a name="objc"></a> - Objc 

```objc
sAPI.host(@"https://xxx.com")
   .method(HTTP_GET)
   .path([NSString stringWithFormat:@"/v1/%lu/users", (unsigned long)roomId])
   .execute().then(^id(NSDictionary *ret){
      // process api ret in main thread
      NSArray *users = ret[@"users"];
      // do business logic ...
      return nil;
   }).catch(^(NSError *error){
      // 错误处理
      NSInteger code = error.code;
      NSString *errorMsg = error.userInfo[sAPIErrorMessageKey]; //errorMsg could be @""
      // do with error info ...
   });

```
### <a name="swift"></a> - Swift

```swift
sAPI.host("https://xxx.com").method(sHTTPMethod.HTTP_GET)
   .path("/v1/\(roomId)/users")
   .execute().__onQueue(DispatchQueue.main, then:{ ret in
        // process api ret in main thread
        // 这里提取数据的类型转换有点繁琐，推荐用 SwiftyJSON 等三方库来处理json对象
        let dic = ret as! [String : AnyObject]
        let users = dic["users"] as! Array<[String : AnyObject]>
        // do business logic ...
        return nil
   }).__onQueue(DispatchQueue.main, catch:{ error in
        // 错误处理
        let error = error as NSError
        let code = error.code
        let errorMsg = error.userInfo[sAPIErrorMessageKey] //errorMsg could be ""
        // do with error info ...
   })
```

### <a name="adaptret"></a> - 自适应服务端返回数据结构

```objc
// sAPI会对服务端返回的原始数据中的"data","error code","error message"三部分信息进行提取，并且对这几部分兼容不同的key，以适配不同的服务端团队
// sAPI内部代码：
    NSInteger errorCode = 0;
    NSString *errorMsg = nil;
    if (response.statusCode != 200) {
        errorCode = response.statusCode > 0 ? response.statusCode : 110; /*110代表请求超时*/
    }
    if ([response.responseObject isKindOfClass:NSDictionary.class]) {
        NSDictionary *body = response.responseObject;
        if (sapi_not_null(body[@"errcode"]) || sapi_not_null(body[@"code"]) || sapi_not_null(body[@"errorno"]) || sapi_not_null(body[@"errCode"])) {
            //code not null
            NSInteger code = sapi_not_null(body[@"errcode"]) ? [body[@"errcode"] integerValue] : sapi_not_null(body[@"code"]) ? [body[@"code"] integerValue] : sapi_not_null(body[@"errorno"]) ?  [body[@"errorno"] integerValue] : [body[@"errCode"] integerValue];
            if (code != 200 && code != 0) {
                errorCode = code;
            }
        }
        errorMsg = sapi_not_null(body[@"errmsg"]) ? body[@"errmsg"] : sapi_not_null(body[@"msg"]) ? body[@"msg"] : sapi_not_null(body[@"errormsg"]) ? body[@"errormsg"] : sapi_not_null(body[@"errMsg"]) ? body[@"errMsg"] : nil;
    }
    if (errorCode > 0) {
        if (completion) {
            completion(response, errorCode, errorMsg);
        }
        return;
    }

    if([response.responseObject isKindOfClass:NSDictionary.class]){
        NSDictionary *body = response.responseObject;
        id data = sapi_not_null(body[@"data"]) ? body[@"data"] : sapi_not_null(body[@"ret"]) ? body[@"ret"] : nil;
        if (data != nil) {
            response = [[sNetworkResponse alloc] initWithResponse:response adpatedObject:data];
        }
    }
    if (completion) {
        completion(response, errorCode, errorMsg);
    }
   // ......
```
## 公共参数

```
对于所有api请求都附带的公共参数，不必每次调用都写一次，sAPI可以配置几个全局的回调(blocks)，在每次发起api请求时都会主动去调用，来获取公共的参数
公共参数包括 commonParams 和 commonHeaders
```

### <a name="commonparams"></a> - commonParams 

```objc
/******************************************************************************************************************************
注意：
1. sAPI.commonParams(...)是一个全局的配置，只需要在app启动时配置一次

2. sAPI.commonParams(...)配置的是一个返回字典的block，而不是字典本身，每次请求api前，这个block都会被调用，用来实时生成公共参数字典

3. 每次api调用都可以链式的指定一个commonParamsUsage，它有两个取值：
   sAPICommonParamsUsagePath：公共参数应用到api请求的querystring中，不论请求用的是不是GET方法都会添加至querystring
   sAPICommonParamsUsageParams：公共参数和api的params合并，不指定encodingType时对于GET请求添加到querystring中，否则添加到请求body的json中
   
4. commonParamUsage不是全局的，而是跟每个api绑定的，默认值为 sAPICommonParamsUsagePath
*******************************************************************************************************************************/

// app启动时配置
sAPI.commonParams(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   return [System1API commonParams];
});

// 一次具体的api调用
sAPI.host(@"https://xxx.com")
   .path(@"/test")
   .params(@{"p1": p1, @"p2": p2})
   .commonParamUsage(sAPICommonParamsUsageParams) //指定公共参数和params合并
   .method(HTTP_POST)
   .execute()
```
### <a name="commonheaders"></a> - commonHeaders

```objc
/******************************************************************************************************************************
注意：
1. sAPI.commonHeaders(...)是一个全局的配置，只需要在app启动时配置一次

2. sAPI.commonHeaders(...)配置的是一个返回字典的block，而不是字典本身，每次请求api前，这个block都会被调用，用来实时生成公共headers字典

3. 每次api调用也可以通过sAPI.headers(...)指定和这次api请求绑定的headers，会和commonHeaders返回的字典合并
*******************************************************************************************************************************/

// app启动时配置
sAPI.commonHeaders(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   return [System1API commonHeaders];
});

// 一次具体的api调用
sAPI.host(@"https://xxx.com")
   .path(@"/auth")
   .headers(@{"SECRET": secret}) // 本次请求绑定的headers
   .method(HTTP_GET)
   .execute()
```

### <a name="tags"></a> - 区分不同子系统

```
如上所述，sAPI.commonParams(...) 和 sAPI.commonHeaders(...) 都是全局的配置，都只能配置一个，如果不同的子系统要求的公共参数不一样怎么办呢？
这可以通过tag来解决，每次具体的api请求都可以链式的指定一个tag，这个tag会作为参数传给 sAPI.commonParams(...) 和 sAPI.commonHeaders(...)配置的block：
```
```objc

// in file AppDelegate.m ...
// ......

// 全局的公共params
sAPI.commonParams(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   if([tag isEqualToString:System1API.tag]){
      // 子系统1
      return [System1API commonParams];
   }
   else if([tag isEqualToString:System2API.tag]){
      // 子系统2
      return [System2API commonParams];
   }
   else {
      // 其他子系统
      return @{};
   }
});

// 全局的公共headers
sAPI.commonHeaders(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   if([tag isEqualToString:System1API.tag]){
      // 子系统1
      return [System1API commonHeaders];
   }
   else if([tag isEqualToString:System2API.tag]){
      // 子系统2
      return [System2API commonHeaders];
   }
   else {
      // 其他子系统
      return @{};
   }
});


// ......
// in file System1API.m（子系统1的所有API都放在这里）

/// 子系统公共逻辑
-(sAPIBuilder*)builder{
    return sAPI.tag(System1API.tag).host(System1API.host);
}

/// 某个具体的API接口
-(FBLPromise<NSDictionary*>*)getUsersInRoom:(NSString *)roomId {
    return self.builder.method(HTTP_GET).path([NSString stringWithFormat:@"/v1/%@/users", roomId]).execute();
}


// ......
// in file System2API.m（子系统2的所有API都放在这里）

/// 子系统公共逻辑
-(sAPIBuilder*)builder{
    return sAPI.tag(System2API.tag).host(System2API.host);
}

/// 某个具体的API接口
-(FBLPromise<NSDictionary*>*)getAudiencesInRoom:(NSUInteger)roomId{
    return self.builder.method(HTTP_GET).path([NSString stringWithFormat:@"/v1/%lu/audiences", (unsigned long)roomId]).execute();
}

```




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.3-0 (2021-09-30)


### Features

* **NoTicket:** setup project ([0576e12](https://github.com/jinsikui/sAPI/commit/0576e120bf3b6759b8a2427f6a6dc77d4223077b))


Source/Classes/Services/sAPI.m
//
//  sAPI.m
//  LiveLib_Example
//
//  Created by DemonY on 2020/7/30.
//  Copyright © 2020 蜻蜓fm. All rights reserved.
//

#import "sAPI.h"

@interface sAPIBaseBuilder ()
@property (nonatomic, copy, nullable, readwrite) SAPISetParamsBlock paramsBlock;
@property (nonatomic, copy, nullable, readwrite) SAPISetHeadersBlock headersBlock;
@property (nonatomic, copy) NSString *mDefaultHost;
@property (nonatomic, assign) sHTTPMethod mDefaultMethod;
@end

@implementation sAPIBaseBuilder

+ (instancetype)shared {
    static sAPIBaseBuilder *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[sAPIBaseBuilder alloc] init];
    });
    return instance;
}

- (void (^)(SAPISetParamsBlock _Nonnull))commonParams {
    return ^(SAPISetParamsBlock commonParamsBlock) {
        self.paramsBlock = commonParamsBlock;
    };
}

- (void (^)(SAPISetHeadersBlock _Nonnull))commonHeaders {
    return ^(SAPISetHeadersBlock commonHeadersBlock) {
        self.headersBlock = commonHeadersBlock;
    };
}

- (void (^)(NSString *defaultHost))defaultHost {
    return ^(NSString *defaultHost) {
        self.mDefaultHost = [defaultHost copy];
    };
}

- (void (^)(sHTTPMethod defaultMethod))defaultMethod {
    return ^(sHTTPMethod defaultMethod) {
        self.mDefaultMethod = defaultMethod;
    };
}

@end

@interface sAPIConvertable () {
    NSDictionary *_mParams;
    NSDictionary *_mHeaders;
    NSString *_mHost;
    NSString *_mPath;
    sHTTPMethod _mMethod;
    sParameterEncoding _mEncodingType;
    sResponseDecncoding _mDecodingType;
    sNetworkSub *_mStubData;
}
@end
@implementation sAPIConvertable

- (instancetype)initWithParams:(NSDictionary *)params
                       headers:(NSDictionary *)headers
                          host:(NSString *)host
                          path:(NSString *)path
                        method:(sHTTPMethod)method
                  encodingType:(sParameterEncoding)encodingType
                  decodingType:(sResponseDecncoding)decodingType
                      stubData:(sNetworkSub*)stubData{
    self = [super init];
    if (self) {
        _mParams = params;
        _mHeaders = headers;
        _mHost = host;
        _mPath = path;
        _mMethod = method;
        _mEncodingType = encodingType;
        _mDecodingType = decodingType;
        _mStubData = stubData;
    }
    return self;
}

#pragma mark - sNetworkConvertable
- (NSString*)baseURL {
    return _mHost;
}

- (NSDictionary<NSString *,NSString *> *)httpHeaders {
    return _mHeaders;
}

- (NSString *)path {
    return _mPath;
}

- (sHTTPMethod)httpMethod {
    return _mMethod;
}

- (NSDictionary *)parameters {
    return _mParams;
}

- (sParameterEncoding)encodingType {
    return _mEncodingType;
}

- (sResponseDecncoding)decodingType {
    return _mDecodingType;
}

- (sNetworkSub*)stubData{
    return _mStubData;
}

- (sNetworkResponse *)adaptResponse:(sNetworkResponse *)networkResponse {
    return [[sNetworkResponse alloc] initWithResponse:networkResponse adpatedObject:networkResponse.responseObject];
}

@end

@interface sAPIBuilder ()

@property (nonatomic, copy, readwrite) NSString *mTag;
@property (nonatomic, strong, readwrite) sNetworkSub *mStubData;
@property (nonatomic, strong, readwrite) NSDictionary *mParams;
@property (nonatomic, strong, readwrite) NSDictionary *mHeaders;
@property (nonatomic, assign, readwrite) sAPICommonParamsUsage mParamUsage;
@property (nonatomic, assign, readwrite) sAPICommonHeadersUsage mHeaderUsage;
@property (nonatomic, copy, readwrite) NSString *mHost;
@property (nonatomic, copy, readwrite) NSString *mPath;
@property (nonatomic, assign, readwrite) sHTTPMethod mMethod;
@property (nonatomic, assign, readwrite) sParameterEncoding mEncodingType;
@property (nonatomic, assign, readwrite) sResponseDecncoding mDecodingType;
@property (nonatomic, copy, nullable, readwrite) SAPIBeforeNetworkCallbackBlock beforeCallback;

@property (strong, nonatomic) sNetworkManager * manager;
@end

@implementation sAPIBuilder

- (instancetype)init {
    self = [super init];
    if (self) {
        self.manager = [sNetworkManager manager];
        self.mEncodingType = sParameterEncodingJSON;
        self.mDecodingType = sResponseDecncodingJSON;
        self.mParamUsage = sAPICommonParamsUsagePath;
        self.mHeaderUsage = sAPICommonHeadersUsageHeaders;
        self.mMethod = [sAPIBaseBuilder shared].mDefaultMethod;
        self.mHost = [sAPIBaseBuilder shared].mDefaultHost;
    }
    return self;
}

- (sAPIBuilder * _Nonnull (^)(NSString * _Nonnull))tag {
    return ^id(NSString *tag) {
        self.mTag = tag;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(sNetworkSub * _Nonnull))stubData {
    return ^id(sNetworkSub *stubData) {
        self.mStubData = stubData;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))params {
    return ^id(NSDictionary *params) {
        self.mParams = params;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))headers {
    return ^id(NSDictionary *headers) {
        self.mHeaders = headers;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(sAPICommonParamsUsage))commonParamsUsage {
    return ^id(sAPICommonParamsUsage usage) {
        self.mParamUsage = usage;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(sAPICommonHeadersUsage))commonHeadersUsage {
    return ^id(sAPICommonHeadersUsage usage) {
        self.mHeaderUsage = usage;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(NSString * _Nonnull))host {
    return ^id(NSString *host) {
        self.mHost = [host copy];
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(NSString * _Nonnull))path {
    return ^id(NSString *path) {
        self.mPath = [path copy];
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(sHTTPMethod))method {
    return ^id(sHTTPMethod method) {
        self.mMethod = method;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(sParameterEncoding))encodingType {
    return ^id(sParameterEncoding encodingType) {
        self.mEncodingType = encodingType;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(sResponseDecncoding))decodingType {
    return ^id(sResponseDecncoding decodingType) {
        self.mDecodingType = decodingType;
        return self;
    };
}

- (sAPIBuilder * _Nonnull (^)(SAPIBeforeNetworkCallbackBlock _Nonnull))beforeNetworkCallback {
    return ^id(SAPIBeforeNetworkCallbackBlock beforeCallback) {
        self.beforeCallback = beforeCallback;
        return self;
    };
}

- (FBLPromise * (^)(void))execute {
    return ^FBLPromise * {
        FBLPromise * (^networkPromise)(void) = ^{
            return FBLPromise.asyncOn(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(FBLPromiseFulfillBlock fulfill, FBLPromiseRejectBlock reject) {
                void (^myCompletion)(sNetworkResponse * _Nonnull, NSInteger, NSString*) = ^(sNetworkResponse * _Nonnull response, NSInteger errorCode, NSString *errorMsg) {
                    if (errorCode != 0) {
                        NSError *error = [NSError errorWithDomain:sAPIErrorDomain code:errorCode userInfo:@{sAPIErrorMessageKey: errorMsg ?: @""}];
                        reject(error);
                    } else {
                        fulfill(response.responseObject);
                    }
                };
                [self.manager request:[self createConvertable] completion:[self getRequestCompletionFromCompletion:myCompletion]];
            });
        };
        /// 是否有在网络请求之前的回调
        if (self.beforeCallback) {
            FBLPromise *beforePromise = self.beforeCallback();
            if (beforePromise) {
                return beforePromise.then(^id _Nullable(id  _Nullable value) {
                    return networkPromise();
                });
            }
        }
        return networkPromise();
    };
}

- (sAPIConvertable *)createConvertable {
    NSMutableDictionary *commonParams = [NSMutableDictionary dictionary];
    if ([sAPIBaseBuilder shared].paramsBlock) {
        commonParams = [[sAPIBaseBuilder shared].paramsBlock(self.mTag) mutableCopy];
    }
    NSMutableDictionary *commonHeaders = [NSMutableDictionary dictionary];
    if ([sAPIBaseBuilder shared].headersBlock) {
        commonHeaders = [[sAPIBaseBuilder shared].headersBlock(self.mTag) mutableCopy];
    }
    /// 参数
    if (self.mParamUsage & sAPICommonParamsUsageParams) {
        [commonParams addEntriesFromDictionary:self.mParams];
        self.mParams = commonParams;
    }
    /// 头
    if (self.mHeaderUsage & sAPICommonHeadersUsageHeaders) {
        [commonHeaders addEntriesFromDictionary:self.mHeaders];
        self.mHeaders = commonHeaders;
    }
    /// 地址
    if (self.mParamUsage & sAPICommonParamsUsagePath) {
        self.mPath = [sAPIHelper mergeToInput:self.mPath queryParams:commonParams];
    }
    
    /// 编码
    sParameterEncoding encodingType = self.mMethod == HTTP_GET ? sParameterEncodingHTTP : self.mEncodingType;
    
    return [[sAPIConvertable alloc] initWithParams:self.mParams
                                           headers:self.mHeaders
                                              host:self.mHost
                                              path:self.mPath
                                            method:self.mMethod
                                      encodingType:encodingType
                                      decodingType:self.mDecodingType
                                          stubData:self.mStubData];
}

- (void(^)(sNetworkResponse * _Nonnull))getRequestCompletionFromCompletion:(void (^)(sNetworkResponse * _Nonnull, NSInteger errorCode, NSString *errorMsg))completion {
    void (^requestCompletion)(sNetworkResponse * _Nonnull) = ^(sNetworkResponse *response) {
        NSInteger errorCode = 0;
        NSString *errorMsg = nil;
        if (response.statusCode != 200) {
            errorCode = response.statusCode > 0 ? response.statusCode : 110; /*110代表请求超时*/
        }
        if ([response.responseObject isKindOfClass:NSDictionary.class]) {
            NSDictionary *body = response.responseObject;
            if (sapi_not_null(body[@"errcode"]) || sapi_not_null(body[@"code"]) || sapi_not_null(body[@"errorno"]) || sapi_not_null(body[@"errCode"])) {
                //code not null
                NSInteger code = sapi_not_null(body[@"errcode"]) ? [body[@"errcode"] integerValue] : sapi_not_null(body[@"code"]) ? [body[@"code"] integerValue] : sapi_not_null(body[@"errorno"]) ?  [body[@"errorno"] integerValue] : [body[@"errCode"] integerValue];
                if (code != 200 && code != 0) {
                    errorCode = code;
                }
            }
            errorMsg = sapi_not_null(body[@"errmsg"]) ? body[@"errmsg"] : sapi_not_null(body[@"msg"]) ? body[@"msg"] : sapi_not_null(body[@"errormsg"]) ? body[@"errormsg"] : sapi_not_null(body[@"errMsg"]) ? body[@"errMsg"] : nil;
        }
        if (errorCode > 0) {
            if (completion) {
                completion(response, errorCode, errorMsg);
            }
            return;
        }
        
        if([response.responseObject isKindOfClass:NSDictionary.class]){
            NSDictionary *body = response.responseObject;
            id data = sapi_not_null(body[@"data"]) ? body[@"data"] : sapi_not_null(body[@"ret"]) ? body[@"ret"] : nil;
            if (data != nil) {
                response = [[sNetworkResponse alloc] initWithResponse:response adpatedObject:data];
            }
        }
        if (completion) {
            completion(response, errorCode, errorMsg);
        }
    };
    return requestCompletion;
}

@end

@implementation sAPI

#pragma mark - Common API Properties

+ (void)buildCommon:(void (^)(sAPIBaseBuilder * _Nonnull baseBuilder))block {
    block([sAPIBaseBuilder shared]);
}

+ (void (^)(SAPISetParamsBlock _Nonnull))commonParams {
    return sAPIBaseBuilder.shared.commonParams;
}

+ (void (^)(SAPISetHeadersBlock _Nonnull))commonHeaders {
    return sAPIBaseBuilder.shared.commonHeaders;
}

+ (void (^)(NSString * _Nonnull))defaultHost {
    return sAPIBaseBuilder.shared.defaultHost;
}

+ (void (^)(sHTTPMethod))defaultMethod {
    return sAPIBaseBuilder.shared.defaultMethod;
}

#pragma mark - Instance API Properties

+ (sAPIBuilder *)build:(void (^)(sAPIBuilder * _Nonnull))block {
    sAPIBuilder *builder = [[sAPIBuilder alloc] init];
    if (block) {
        block(builder);
    }
    return builder;
}

+ (sAPIBuilder * _Nonnull (^)(NSString * _Nonnull))tag {
    return [[[sAPIBuilder alloc] init] tag];
}

+ (sAPIBuilder * _Nonnull (^)(sNetworkSub * _Nonnull))stubData {
    return [[[sAPIBuilder alloc] init] stubData];
}

+ (sAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))params {
    return [[[sAPIBuilder alloc] init] params];
}

+ (sAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))headers {
    return [[[sAPIBuilder alloc] init] headers];
}

+ (sAPIBuilder * _Nonnull (^)(sAPICommonParamsUsage))commonParamsUsage {
    return [[[sAPIBuilder alloc] init] commonParamsUsage];
}

+ (sAPIBuilder * _Nonnull (^)(sAPICommonHeadersUsage))commonHeadersUsage {
    return [[[sAPIBuilder alloc] init] commonHeadersUsage];
}

+ (sAPIBuilder * _Nonnull (^)(NSString * _Nonnull))host {
    return [[[sAPIBuilder alloc] init] host];
}

+ (sAPIBuilder * _Nonnull (^)(NSString * _Nonnull))path {
    return [[[sAPIBuilder alloc] init] path];
}

+ (sAPIBuilder * _Nonnull (^)(sHTTPMethod))method {
    return [[[sAPIBuilder alloc] init] method];
}

+ (sAPIBuilder * _Nonnull (^)(sParameterEncoding))encodingType {
    return [[[sAPIBuilder alloc] init] encodingType];
}

+ (sAPIBuilder * _Nonnull (^)(SAPIBeforeNetworkCallbackBlock _Nonnull))beforeNetworkCallback {
    return [[[sAPIBuilder alloc] init] beforeNetworkCallback];
}

@end
